draft-ietf-tls-esni.md

```diff
@@ -388,52 +423,59 @@ ClientHelloOuter, and a digest of the complete ClientHelloInner.
+duplicated in ClientHelloOuter. To do so, the client computes a hash of the
 entire ClientHelloInner message with the same hash as for the KDF used to
-encrypt ClientHelloInner. Then, the client removes and and replaces extensions
-from ClientHelloInner with a single "outer_extensions" extension. The list of
-outer_extensions include those which were removed from ClientHelloInner, in the
-order in which they were removed. The hash contains the full ClientHelloInner
-hash H computed above.
-
-This process is reversed by client-facing servers upon receipt. Specifically,
-the server replaces the "outer_extensions" with extensions contained in
-ClientHelloOuter. The server then computes a hash H' of the reconstructed
-ClientHelloInner. If H' does not equal OuterExtensions.hash, the server aborts
-the connection with an "illegal_parameter" alert.
+encrypt ClientHelloInner. Then, the client removes and replaces extensions from
+ClientHelloInner with a single "outer_extensions" extension. The list of
+`outer_extensions` include those which were removed from ClientHelloInner, in
+the order in which they were removed. The hash contains the full
+ClientHelloInner hash computed above.
+
+This process is reversed by client-facing server. Specifically,
+the server replaces the `outer_extensions` with extensions contained in
+ClientHelloOuter. The server then computes a hash of the reconstructed
+ClientHelloInner. If the hash does not equal OuterExtensions.hash, the server
+aborts the connection with an "illegal_parameter" alert.
+
+[[TODO: What does the hash accomplish? It seems superfluous to me.]]
 
 Clients SHOULD only use this mechanism for extensions which are large. All other
 extensions SHOULD appear in both ClientHelloInner and ClientHelloOuter even if
-they have identical values.
+they have identical values. [[TODO: Why the second SHOULD? If I were
+implementing this, I would shallow-copy all redundant extensions, except maybe
+those that have 0-length payloads.]]
 
 # Client Behavior {#client-behavior}
 
 ## Sending an encrypted ClientHello {#send-ech}
 
 In order to send an encrypted ClientHello, the client first determines if it
-supports the server's chosen KEM, as identified by ECHConfig.kem_id. If one is
-supported, the client MUST select an appropriate HpkeCipherSuite from the list
+supports the server's chosen KEM, as identified by `ECHConfig.kem_id`. If one is
+supported, the client MUST select an appropriate `HpkeCipherSuite` from the list
 of suites offered by the server. If the client does not support the
 corresponding KEM or is unable to select an appropriate group or
-HpkeCipherSuite, it SHOULD ignore that ECHConfig value and MAY attempt to use
-another value provided by the server. The client MUST NOT send ECH using HPKE
-algorithms not advertised by the server.
+`HpkeCipherSuite`, it SHOULD ignore that `ECHConfig` value and MAY attempt to
+use another value provided by the server. [[TODO: "appropriate group": Isn't
```

Comment:
I'm not sure I understand the question here. I don't think the text suggests that the group is part of the ciphersuite. We can probably s/"an appropriate group or"/"a"

Reply:
That works for me.
Done.